test/integration/install.py

```diff
@@ -27,20 +31,44 @@ def __init__(self, path):
 			print("WARNING: bitness misalignment between interpreter (%s) and driver (%s): testing will likely fail" %
 					(bitness_py, bitness_driver))
 
+		self._driver_name = "%s (%sbit)" % (DRIVER_BASE_NAME, bitness_driver)
+
 	def _install_driver_win(self, ephemeral):
+		# remove any old driver (of tested bitness)
+		name_filter = "name = '%s'" % self._driver_name
+		with psutil.Popen(["wmic", "/INTERACTIVE:OFF", "product", "where", name_filter, "call", "uninstall"],
+				stdout=PIPE, stderr=PIPE, universal_newlines=True) as p:
+			try:
+				p.wait(WMIC_UNINSTALL_TIMOUT)
+			except psutil.TimeoutExpired as e:
+				try: p.kill()
+				except: pass
+				raise Exception("wmic uninstallation killed after %s seconds" % WMIC_UNINSTALL_TIMOUT)
+
+			assert(p.returncode is not None)
+			out, err = p.communicate()
+			if (out): print(out)
+			if (err): print(err)
+			if p.returncode:
+				print("ERROR: driver wmic-uninstallation failed.")
+			else:
+				print("INFO: an old driver was %s." % ("uninstalled" if ("ReturnValue = 0" in out) else "not found"))
+
+
+		# get a file name to log the installation into
 		(fd, log_name) = tempfile.mkstemp(suffix="-installer.log")
 		os.close(fd)
 		if ephemeral:
-			atexit.register(shutil.rmtree, log_name, ignore_errors=True)
-
+			atexit.register(os.remove, log_name)
+		# install the new driver
 		with psutil.Popen(["msiexec.exe", "/i", self._driver_path, "/norestart", "/quiet", "/l*vx", log_name]) as p:
-			waiting_since = time.time()
-			while p.poll() is None:
-				time.sleep(.3)
-				if Elasticsearch.TERM_TIMEOUT < time.time() - waiting_since:
-					try: p.kill()
-					except: pass
-					raise Exception("installer killed after %s seconds" % Elasticsearch.TERM_TIMEOUT)
+			try:
+				p.wait(Elasticsearch.TERM_TIMEOUT)
+			except psutil.TimeoutExpired as e:
+				try: p.kill()
```

Comment:
Windows installer tries to roll back what it has done if it is stopped gracefully while running.  But `p.kill()` will terminate the process in a way that doesn't give it a chance to do any cleanup.  This could leave the registry in a state where the product Windows installer was working on can neither be installed nor uninstalled without errors.  If this ever happened the simplest way to recover a build VM would be to scrap it and create a new VM.  It might be worth adding something to this effect in the exception message.  Effectively, if this error occurs the build VM should not be reused as it will cause more time to be wasted debugging the following installation failure.

Reply:
Thanks a good point, thanks.
I've added a warning line, but also removed the signaling, for both un- and installation, since the testing application would eventually abort anyways.
(The OS shutdown might hopefully end the processes more gracefully.)